File: src/interfaces/contract.ts
import type { ClientOptions } from "@stellar/stellar-sdk/contract";
import { CarStatus } from "./car-status.ts";

export interface IBaseContractClient {
  readonly options: ClientOptions;
  toXDR(): string;
}

export interface IRentACarContract extends IBaseContractClient {
  __constructor: ({
    admin,
    token,
  }: {
    admin: string;
    token: string;
  }) => Promise<this>;

  set_admin_commission: ({
    commission,
  }: {
    commission: number;
  }) => Promise<this>;
  get_admin_commission: () => Promise<number>;
  get_admin_balance: () => Promise<number>;

  add_car: ({
    owner,
    price_per_day,
  }: {
    owner: string;
    price_per_day: number;
  }) => Promise<this>;

  get_car_status: ({ owner }: { owner: string }) => Promise<CarStatus>;

  rental: ({
    renter,
    owner,
    total_days_to_rent,
  }: {
    renter: string;
    owner: string;
    total_days_to_rent: number;
  }) => Promise<this>;

  return_car: ({
    renter,
    owner,
  }: {
    renter: string;
    owner: string;
  }) => Promise<this>;

  remove_car: ({ owner }: { owner: string }) => Promise<this>;

  payout_owner: ({
    owner,
    amount,
  }: {
    owner: string;
    amount: number;
  }) => Promise<this>;

  payout_admin: ({ amount }: { amount: number }) => Promise<this>;
}

File: contracts/rent-a-car/src/interface/contract.rs
use soroban_sdk::{Address, Env};
use crate::storage::types::{CarStatus, Error};

pub trait RentACarContractTrait {
    fn __constructor(env: &Env, admin: Address, token: Address) -> Result<(), Error>;
    fn set_admin_commission(env: &Env, commission: i128) -> Result<(), Error>;
    fn get_admin_commission(env: &Env) -> i128;
    fn add_car(env: &Env, owner: Address, price_per_day: i128) -> Result<(), Error>;
    fn get_car_status(env: &Env, owner: Address) -> Result<CarStatus, Error>;
    fn rental(env: &Env, renter: Address, owner: Address, total_days_to_rent: u32) -> Result<(), Error>;
    fn return_car(env: &Env, renter: Address, owner: Address) -> Result<(), Error>;
    fn remove_car(env: &Env, owner: Address) -> Result<(), Error>;
    fn payout_owner(env: &Env, owner: Address, amount: i128) -> Result<(), Error>;
    fn payout_admin(env: &Env, amount: i128) -> Result<(), Error>;
    fn get_admin_balance(env: &Env) -> i128;
}